### PR TITLE
Build system

### LINKED ISSUE #1: Broken relative paths in makefiles
The `../../../FDTD` path parts in the following lines

https://github.com/UCL/isotropic-finite-difference-time-domain/blob/b6136cb6f16b270425b78b530545bf8e60df498e/iterater/MakefileOMP_PSTD_UCL2#L76

https://github.com/UCL/isotropic-finite-difference-time-domain/blob/b6136cb6f16b270425b78b530545bf8e60df498e/iterater/MakefileOMP_FDTD_UCL2#L19

https://github.com/UCL/isotropic-finite-difference-time-domain/blob/b6136cb6f16b270425b78b530545bf8e60df498e/iterater/MakefileOMP_FDTD_UCL2#L73-L74

should I think just be `..`

The following lines also refer to directories outside of the project root but are inside a conditional which is not entered when running the file as is

https://github.com/UCL/isotropic-finite-difference-time-domain/blob/b6136cb6f16b270425b78b530545bf8e60df498e/iterater/MakefileOMP_PSTD_UCL2#L39-L40
https://github.com/UCL/isotropic-finite-difference-time-domain/blob/b6136cb6f16b270425b78b530545bf8e60df498e/iterater/MakefileOMP_FDTD_UCL2#L37-L38

### PR DESCRIPTION
**Pull request by t-young31**
Migrate build system from standalone makefiles to [CMake](https://cmake.org/), which should enable more portable compilation. 

Closes #1 #4 #6 #9 

***
### Notes

- Removes option and code associated with using non-MATLAB i/o

***
### Usage

If both MATLAB and fftw can be found with their associated cmake `find_package` functions tdms can be compiled with

```bash
cd tdms
rm -rf build; mkdir build; cd build
cmake .. && make install
```

without write access to the default `CMAKE_INSTALL_PREFIX` (_/usr/local/_ on unix) a local directory can be specified e.g.

```bash
cmake .. -DCMAKE_INSTALL_PREFIX=$HOME/.local/
``` 

where the executable will be located in _~/.local/bin_.

---
My M1 mac needs the following cmake flags, due to MATLAB only providing x86 libs so openmp and libc++ need to be also x86, and so specified explicitly:
```bash
-DMatlab_ROOT_DIR=/Applications/MATLAB_R2022a.app -DFFTW_ROOT=/Users/tom/.local/fftw3_x86/ -DCMAKE_CXX_COMPILER=/Users/tom/.local/homebrew/opt/llvm/bin/clang++ -DOMP_ROOT=/Users/tom/.local/homebrew/opt/llvm/ -DCXX_ROOT=/Users/tom/.local/homebrew/opt/llvm
```

On [Myriad](https://www.rc.ucl.ac.uk/docs/Clusters/Myriad/) a few modules need to be loaded:
```bash
module load beta-modules; module swap gcc-libs/9.2.0; module swap compilers/gnu/9.2.0; module load xorg-utils matlab/full/r2019b/9.7 fftw/3.3.6-pl2/gnu-4.9.2
```

***
### TODO

- [x] Finish triaging files
- [ ] Test local compilation: @giordano @prmunro
- [x] Test binary reproduces test output (dependency: #2)
- [x] Update README to reflect changes
- [x] Enable FD code compilation
- [x] Add ending newlines
- [x] Reduce default parallelism (use $OMP_NUM_THREADS)

***
### Questions

- Shall we delete the [triage/](https://github.com/t-young31/ifdtd/tree/build_system/triage) directory? (noting the files aren't 'gone' and can always be copied from a previous commit)
- Should anything else be added to the [README](https://github.com/t-young31/ifdtd/blob/build_system/README.md)? (feel free to edit and push to this branch if that's easier)
- Is there a FDTD script example that could be used for testing?

### CODE REVIEW

File: tdms/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    set(DFLAG "-DFDFLAG")
+endif()
+
+
+# Matlab ----------------------------------------------------------------------
+find_package(Matlab REQUIRED MAT_LIBRARY)
+include_directories(${Matlab_INCLUDE_DIRS})
+
+# Currently matlab only has x84 libraries on Mac (no arm)
+set(CMAKE_OSX_ARCHITECTURES "x86_64")
```
Comment (giordano) [inline tdms/CMakeLists.txt:38]: Sounds like there is already a beta version native for Apple Silicon: https://blogs.mathworks.com/matlab/2022/05/05/exploring-the-matlab-beta-for-native-apple-silicon/.  Maybe you can try this one out?  This should simplify the build system a bit.

File: tdms/cmake/FindFFTW.cmake [hunk 2]
```diff
@@ -0,0 +1,97 @@
+# - Find the FFTW library
+#
+# Adapted from: https://github.com/egpbos/findFFTW/blob/master/FindFFTW.cmake
```
[the remark this answers is not in the document]
Comment (t-young31) [reply tdms/cmake/FindFFTW.cmake:3]: Have elaborated, and it does but I'm not sure they're sufficient to _find_ fftw

File: tdms/CMakeLists.txt (hunk 1 above)
Comment (t-young31) [reply tdms/CMakeLists.txt:38]: I've spent a moment trying and [FindMatlab.cmake](https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/FindMatlab.cmake#L1741) is hard coded for intel, and I'm disinclined to copy it into this project and modify or edit the upstream for something that is in beta 😕

**Review by giordano — COMMENTED** [review]
After changing the required version of cmake to 3.21 to at least be able to run the build on Myriad, I had to load the following modules (some of them are different from what was suggested above, I'm loading also cmake, and using a more recent Matlab):
```
module purge
module load beta-modules
module load gcc-libs/9.2.0 compilers/gnu/9.2.0 xorg-utils matlab/full/r2021a/9.10 fftw/3.3.6-pl2/gnu-4.9.2 cmake/3.21.1
```
but then configuration failed to find some Matlab components, although Matlab _is_ found:
```console
$ cmake ../tdms/
-- The C compiler identification is GNU 9.2.0
-- The CXX compiler identification is GNU 9.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /shared/ucl/apps/gcc/9.2.0/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /shared/ucl/apps/gcc/9.2.0/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Matlab: /lustre/shared/ucl/apps/Matlab/R2021a/full/extern/include (found version "9.10") found components: MAT_LIBRARY 
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.27.1") 
-- Found FFTW: /shared/ucl/apps/fftw/3.3.6-pl2/gnu-4.9.2/include   
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Configuring done
CMake Error at CMakeLists.txt:60 (add_executable):
  Target "tdms" links to target "Matlab::mex" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?


CMake Error at CMakeLists.txt:60 (add_executable):
  Target "tdms" links to target "Matlab::mx" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?


CMake Error at CMakeLists.txt:60 (add_executable):
  Target "tdms" links to target "Matlab::mat" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?


-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

File: tdms/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        src/tdms_iterator.cpp
+        src/mesh_base.cpp
+        src/numerical_derivative.cpp
+        matlabio/matlabio.cpp
+        )
+
+target_link_libraries(tdms
+        Matlab::mex
+        Matlab::mx
+        Matlab::mat
```
Comment (giordano) [inline tdms/CMakeLists.txt:75]: With this
```suggestion
        ${Matlab_MEX_LIBRARY}
        ${Matlab_MX_LIBRARY}
        ${Matlab_MAT_LIBRARY}
```
I can configure and build successfully the project.

File: tdms/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,81 @@
+project(tdms)
+
+# General setup ---------------------------------------------------------------
+cmake_minimum_required(VERSION 3.22)
```
[the remark this answers is not in the document]
Comment (giordano) [reply tdms/CMakeLists.txt:4]: Oh, it sounds like the imported targets are [new in CMake 3.22](https://cmake.org/cmake/help/latest/module/FindMatlab.html#imported-targets)?  But still, that's a pretty new version, hard to expect to be available on a cluster.

[the remark this answers is not in the document]
Comment (t-young31) [reply tdms/CMakeLists.txt]: > the imported targets are [new in CMake 3.22](https://cmake.org/cmake/help/latest/module/FindMatlab.html#imported-targets)

bingo. Agree that requiring a slightly older version of CMake is probably sensible though – have downgraded appropriately. (as an aside: I always install miniconda on a cluster to get a consistent env. Python/git/cmake etc.)

[the remark this answers is not in the document]
Comment (giordano) [reply tdms/CMakeLists.txt]: With the latest version I get
```console
$ cmake ../tdms/ 
-- The C compiler identification is Intel 18.0.3.20180410
-- The CXX compiler identification is Intel 18.0.3.20180410
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /shared/ucl/apps/intel/2018.Update3/compilers_and_libraries_2018.3.222/linux/bin/intel64/icc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /shared/ucl/apps/intel/2018.Update3/compilers_and_libraries_2018.3.222/linux/bin/intel64/icpc - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at /lustre/shared/ucl/apps/cmake/3.21.1/gnu-4.9.2/share/cmake-3.21/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Matlab (missing: Matlab_INCLUDE_DIRS Matlab_MEX_LIBRARY
  Matlab_MEX_EXTENSION Matlab_ROOT_DIR Matlab_MX_LIBRARY MAT_LIBRARY) (found
  version "NOTFOUND")
Call Stack (most recent call first):
  /lustre/shared/ucl/apps/cmake/3.21.1/gnu-4.9.2/share/cmake-3.21/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /lustre/shared/ucl/apps/cmake/3.21.1/gnu-4.9.2/share/cmake-3.21/Modules/FindMatlab.cmake:1868 (find_package_handle_standard_args)
  CMakeLists.txt:34 (find_package)


-- Configuring incomplete, errors occurred!
```

**Review by giordano — APPROVED** [review]
Working for me on Myriad now.

File: tdms/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,79 @@
+project(tdms)
+
+# General setup ---------------------------------------------------------------
+cmake_minimum_required(VERSION 3.21)
+
+# Allow RPATH on mac
+set(CMAKE_MACOSX_RPATH 1)
```
[the remark this answers is not in the document]
Comment (t-young31) [reply tdms/CMakeLists.txt:7]: > allowing [plenty of truthy-falsy values](https://cmake.org/cmake/help/latest/command/if.html#basic-expressions)

well that is awful – thanks for the suggestion 👍🏼

**Pull request merged**